### ISSUE TITLE
Installed Windows 10 OVMF SMP guest crashes when SMM is enabled/required (Bugzilla Bug 85)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#6608]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 85
Date: 2016-07-21T16:12:10+00:00
From: @lersek
To:   @lersek
CC:   @lersek

See also:     https://github.com/tianocore/edk2/issues/91
Last updated: 2016-07-21T16:31:30+00:00

**Comment by tianocore-issues** [comment]
## Comment 321
Date: 2016-07-21 16:12:10 +0000
From: @lersek

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): @lersek

webczat commented on May 21

Hello.

It seems that it is not possible to install windows10 on qemu virtual machine when the machine has more than one virtual core and uses ovmf as a firmware. Of course kv is enabled.
I know there was a similar issue reported here where ovmf itself failed to boot with error like &quot;entry failed: 80000021). I currently have no logs and the number of the issue where that one was reported but the error is identical, of course except that the register dump below the error is different.
The difference between the previously reported issue and this one is that it does not happen during firmware boot. I am able to successfully boot the windows installation media, and I am able to install the system. The problem is that during the first startup/oob experience, it goes normally, then suddenly the machine halts, dumping the above error. I would guess that if you tried to boot an already installed and configured system, the same thing would happen.
I have verified that the system can be installed and booted when there is only one virtual core. It also works when using seabios instead of ovmf, that&apos;s the reasoning behind reporting it here.

My hardware and system information:
4 gigabytes of RAM memory
processor model name : Intel(R) Celeron(R) CPU N2820 @ 2.13GHz
linux kernel version: 4.5.4 (archlinux, so almost vanilla)
ovmf version: latest git, or almost latest, from the same week the issue was reported in
This is the machine bootup script:

#!/bin/sh
exec /usr/bin/qemu-system-x86_64 \
-nodefaults \
-machine q35,accel=kvm,smm=on -smp 2,maxcpus=4,sockets=1,cores=2,threads=1 -m 2048 \
-device qxl-vga -device ich9-intel-hda -device hda-micro \
-netdev vde,id=net -device virtio-net-pci,netdev=net,disable-modern=false \
-device virtio-rng-pci,disable-modern=false -device virtio-balloon-pci,disable-modern=false -chardev stdio,id=mux,mux=on -mon chardev=mux,mode=readline -device isa-serial,chardev=mux \
-display sdl -name winvm -device pvpanic -device ib700 \
-device nec-usb-xhci -drive id=cd,if=none,media=cdrom -device ide-cd,drive=cd,bus=ide.0 \
-drive id=cd2,if=none,media=cdrom -device ide-cd,bus=ide.1,drive=cd2 \
-drive id=disk,file=disk.img,if=none,media=disk -device virtio-blk-pci,drive=disk,disable-modern=false \
-cpu host \
-global driver=cfi.pflash01,property=secure,value=on \
-drive if=pflash,format=raw,file=firmware.bin -drive if=pflash,file=vars.bin,format=raw \
$@

**Comment by tianocore-issues** [comment]
## Comment 322
Date: 2016-07-21 16:13:16 +0000
From: @lersek

webczat commented on May 21

I was previously referring to issue #80 [--&gt;bug 83]

**Comment by tianocore-issues** [comment]
## Comment 323
Date: 2016-07-21 16:13:46 +0000
From: @lersek

lersek commented on May 23 • edited

I understand it doesn&apos;t reproduce with SeaBIOS. I still think this is more related to KVM than to OVMF (given that the firmware itself boots, and the Windows 10 installation succeeds too).

Am I right to think that this is yet another regression that was brought about by a host kernel update? (This is the first report I&apos;m seeing against host kernel 4.5.*.) If so, can you please bisect the host kernel? Maybe even capture a KVM trace &lt;http://www.linux-kvm.org/page/Tracing&gt; please? Equipped with those, if this is a host kernel regression, then it should be reported to the kernel bugzilla. Thanks.

**Comment by tianocore-issues** [comment]
## Comment 324
Date: 2016-07-21 16:13:56 +0000
From: @lersek

lersek commented on May 23

Alternatively, if you are convinced this is due to an OVMF package upgrade, can you please bisect OVMF? Thanks.

**Comment by tianocore-issues** [comment]
## Comment 325
Date: 2016-07-21 16:14:03 +0000
From: @lersek

lersek added the OVMF label on May 23

**Comment by tianocore-issues** [comment]
## Comment 326
Date: 2016-07-21 16:14:14 +0000
From: @lersek

lersek commented on May 23

(Temporarily labeling this issue &quot;OVMF&quot; just because no issue should be unlabeled, and not because I&apos;m confirming the problem is with OVMF.)

**Comment by tianocore-issues** [comment]
## Comment 327
Date: 2016-07-21 16:14:30 +0000
From: @lersek

webczat commented on May 23

I am sure this happens also with kernel 4.4.
Because it only happens with ovmf, and only with smp vm, ... then even
if it is not an ovmf bug, ovmf or windows uefi operation causes it. I
believe this bug is related to launching the AP processor, or something
like that.... I mean it *may* be possible you have more knowledge than
me for purposes of tracing this one.
As for bisecting, it may be quite hard for me as on this laptop building
a kernel may take ... really long, unless I can find binaries? Do you
have *any* suggestion as to what kernel version may be the good one, to
have something to test at first?

**Comment by tianocore-issues** [comment]
## Comment 328
Date: 2016-07-21 16:18:12 +0000
From: @lersek

lersek commented on May 23

My impression has been that the latest v4.4 stable release(s) of the host
kernel made things work. Please see issue #21 [--&gt; bug 73] and the
associated kernel BZ.

There has been one exception to that: issue #80 [--&gt;bug 83] (which we closed
in this tracker after relocating the problem to another kernel BZ
&lt;https://bugzilla.kernel.org/show_bug.cgi?id=116731&gt;).

Given your report, I immediately thought that it was a duplicate of either
issue #21 [--&gt; bug 73] or issue #80 [--&gt;bug 83] (i.e., kernel BZs 107561 or
116731). However, since you mentioned issue #80 yourself, I thought you had
read through it (including kernel BZ 116731 as well), and determined that
your symptoms were due to a different root cause. Is that the case? Did you
test the host kernel patches recommended by @rkrcmar in kernel BZ 116731?

&gt; even if it is not an ovmf bug, ovmf or windows uefi operation causes it

I&apos;m asking you to isolate the problematic component. Problems like this have
been reported at least twice thus far, with both cases ending up in the
kernel bugzilla. Personally I can&apos;t even reproduce such problems any longer
(after the KVM fix for issue #21 [--&gt; bug 73]), let alone debug them.

Here&apos;s a list of steps:

* please see if your problem goes away when running the latest v4.4.* stable
  release (4.4.11 apparently) on the host

* if it doesn&apos;t, please see if the discussion in issue #80 [--&gt;bug 83]
  applies to your hardware (i.e., no EPT). If it applies, please work with
  @rkrcmar in kernel BZ 116731.

* If none of the above apply, I don&apos;t know how I could help, without having
  access to your hardware. (Either way, I&apos;d prefer if this problem was
  tracked in the kernel bugzilla, given the history above, perhaps as a
  separate KVM report.)

**Comment by tianocore-issues** [comment]
## Comment 329
Date: 2016-07-21 16:19:19 +0000
From: @lersek

webczat commented on May 23

Symptoms of this issue are the same (probably) as issue #80 [--&gt;bug 83]. but, I DO have EPT in my cpu, and unrestricted guest mode, and almost everything, so it is not the same issue.

**Comment by tianocore-issues** [comment]
## Comment 330
Date: 2016-07-21 16:19:47 +0000
From: @lersek

webczat commented on May 23

It is really not issue #21 [--&gt; bug 73], the whole error is really different from mine.

**Comment by tianocore-issues** [comment]
## Comment 331
Date: 2016-07-21 16:20:09 +0000
From: @lersek

lersek changed the title from &quot;unable to install windows10 with ovmf when smp is used&quot; to &quot;Installed Windows 10 OVMF SMP guest crashes when SMM is enabled/required&quot; on May 23

**Comment by tianocore-issues** [comment]
## Comment 332
Date: 2016-07-21 16:20:38 +0000
From: @lersek

lersek commented on May 23

Thanks for checking.

Actually, now I can see that you are running QEMU with SMM enabled, and with SMM-only access to the pflash chip being enforced. (Also, you are apparently running a self-built OVMF binary -- can you please state your OVMF build options?)

SMM makes a huge difference, and I wish you had pointed it out explicitly in the report. :(

If I had realized your use of SMM, I wouldn&apos;t have mentioned those other reports at all. I have occasionally seen such issues with SMM enabled, and I even reported it once (as yet internally) to @bonzini, after some initial (superficial) investigation myself. I believe Paolo has it on his TODO list to check it out.

For me the circumstances that (rarely) trigger similar problems are not a single SMP guest. Instead, for me it shows up occasionally when I run an SMM guest in parallel with another guest (SMM or not doesn&apos;t matter in the second guest).

I recommend to report this issue to the kernel bugzilla (Product=Virtualization, Component=kvm), with as much information about your hardware, host kernel, guest configuration, and OVMF binary as possible. Please include an ignore_loglevel dmesg from the host kernel (from after the guest crash), the QEMU error report, and a DEBUG_VERBOSE enabled OVMF log too.

I obviously cannot claim for sure that it&apos;s a KVM (and not an OVMF) bug, but at this point KVM seems more likely, and the KVM maintainers are definitely better equipped to debug it at this stage. Thanks.

**Comment by tianocore-issues** [comment]
## Comment 333
Date: 2016-07-21 16:20:53 +0000
From: @lersek

webczat commented on May 23

Yes, I had smm enabled. I built ovmf with ipv6, http boot and secureboot and smm.

**Comment by tianocore-issues** [comment]
## Comment 334
Date: 2016-07-21 16:21:04 +0000
From: @lersek

webczat commented on May 23


Also, my issue is always reproducible. it is not showing up sometimes,
but always at the same moment. I may try building smm and nonsmm ovmf,
then testing with both, both using and not using smp. I will probably
first fully install the os on a working configuration, that is
uniprocessor vm with smm enabled. and try if a fully installed os shows
the same issue when booting

**Comment by tianocore-issues** [comment]
## Comment 335
Date: 2016-07-21 16:21:14 +0000
From: @lersek

bonzini commented on May 23

Can you send me your OVMF binary (or the commit id + command line you built it from)? Also it can be useful to know whether you can reproduce it with e.g. Windows 7. Thanks!

**Comment by tianocore-issues** [comment]
## Comment 336
Date: 2016-07-21 16:21:33 +0000
From: @lersek

lersek commented on May 23

@webczat

&gt; Also, my issue is always reproducible. it is not showing up sometimes, but
&gt; always at the same moment

That sounds (comparatively) great, actually -- the issue I&apos;ve occasionally encountered is very hard to tie to any guest actions. A deterministic reproducer is always welcome. If you can provide further environment details (including reproducibility with a Windows 7 guest, as Paolo suggests), it might help further.

@bonzini: thank you for chiming in here!

**Comment by tianocore-issues** [comment]
## Comment 337
Date: 2016-07-21 16:21:56 +0000
From: @lersek

webczat commented on May 23

I will now try to build few versions of ovmf, with and without smm. One
more detail: I was building from OvmfPkgIa32X64.dsc, not pure X64.
I do not have any windows7 thing at hand at the moment. I will now test
windows10 with different ovmf configurations to check if smm may be the
cause, or some interaction with smm on the second processor, or
whatever. I will tell you if any of those ovmf configurations will work.
if not, I may then try windows7. I hope it supports uefi mode.

**Comment by tianocore-issues** [comment]
## Comment 338
Date: 2016-07-21 16:22:07 +0000
From: @lersek

webczat commented on May 23

I am not sure what I am doing wrong, but now I am unable to run smm ovmf
at all. some bug introduced in a commit? I pulled the repository today
and it seems to me that it does not work, the version without smm works.
Yes, I ensured that I enabled smm in -machine, and secured my pflash.
Either I forgot about something or there is a new bug.

**Comment by tianocore-issues** [comment]
## Comment 339
Date: 2016-07-21 16:22:18 +0000
From: @lersek

webczat commented on May 23

the above affects build using OvmfPkgX64.dsc...

**Comment by tianocore-issues** [comment]
## Comment 340
Date: 2016-07-21 16:22:55 +0000
From: @lersek

lersek commented on May 23

@webczat

The OvmfPkgX64.dsc build of OVMF does not run if both of the following conditions apply:

* S3 support is enabled on the QEMU command line (which is the default), and
* -D SMM_REQUIRE is passed to the build.

The reason is that the S3 Resume PEIM does not support SMM-based S3 resume when the PEI phase is 64-bit.

Please refer to the S3Verification() function in OvmfPkg/PlatformPei/Platform.c. When these circumstances are recognized, the function prints an error message to the QEMU debug port (the same place where all other OVMF messages go):

&gt; S3Verification: S3Resume2Pei doesn&apos;t support X64 PEI + SMM yet.
&gt; S3Verification: Please disable S3 on the QEMU command line (see the README),
&gt; S3Verification: or build OVMF with &quot;OvmfPkgIa32X64.dsc&quot;.

In your report you mention &quot;I currently have no logs&quot;, and also on your QEMU command line I see no isa-debugcon device / no -debugcon switch. A bit later I asked for an OVMF log specifically. -debugcon is documented in the OvmfPkg/README file as well.

I take it if you had not thrown away the OVMF log, you would have seen the above error message, with the clear instructions it gives. Pray tell, are we wasting each other&apos;s time?

**Comment by tianocore-issues** [comment]
## Comment 341
Date: 2016-07-21 16:23:08 +0000
From: @lersek

webczat commented on May 23

well, I was wondering if it is something like that, but did not remember
how to enable debugcon, and where the instructions were. thanks, then I
only have to test one smm and one non smm builds

**Comment by tianocore-issues** [comment]
## Comment 342
Date: 2016-07-21 16:23:46 +0000
From: @lersek


webczat commented on May 24
I have performed some tests. I tried to install windows10 professional
with two ovmf builds:
one is the x64 one with smm disabled, one is the IA32X64 one with smm
enabled.
Results: the one with smm enabled crashes between first and second
restart after installation, but the non smm ovmf does not crash.
Also, this issue seems not to affect all cases, as if I would have an
already installed and configured win10, it seems it would boot in both
cases.

And here is the actual error:
KVM: entry failed, hardware error 0x80000021



If you&apos;re running a guest on an Intel machine without unrestricted mode

support, the failure can be most likely due to the guest entering an
invalid
state for Intel VT. For example, the guest maybe running in big real
mode
 pionowy pasek przewijania 705 procent.
which is not supported on less recent Intel processors.



EAX=00000001 EBX=00491e8c ECX=966f0000 EDX=000c00e1

ESI=9448c6a0 EDI=c159a180 EBP=9448c410 ESP=9448c310

EIP=00008000 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=1 HLT=0

ES =0000 00000000 ffffffff 00809300

CS =c000 7ffc0000 ffffffff 00809300

SS =0000 00000000 ffffffff 00809300

DS =0000 00000000 ffffffff 00809300

FS =0000 00000000 ffffffff 00809300

GS =0000 00000000 ffffffff 00809300

LDT=0000 00000000 000fffff 00000000

TR =0040 c3f2b070 00000067 00008b00

GDT=     c3f2a000 0000006f

IDT=     00000000 00000000

CR0=00050032 CR2=41b66000 CR3=5d948000 CR4=00000000

DR0=0000000000000000 DR1=0000000000000000 DR2=0000000000000000
DR3=0000000000000
000

DR6=00000000ffff0ff0 DR7=0000000000000400

EFER=0000000000000000
Code=ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff &lt;ff&gt; ff
ff ff f
f ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff

**Comment by tianocore-issues** [comment]
## Comment 343
Date: 2016-07-21 16:24:28 +0000
From: @lersek

bonzini commented on May 24

I cannot reproduce this with Linux 4.7-rc1 and the following command line:

&gt; qemu-system-x86_64 -nodefaults --enable-kvm -machine q35,smm=on -smp 2 -m 2048 \
&gt;      -debugcon mon:stdio -vnc :0 -cpu Haswell-noTSX,enforce \
&gt;      -global driver=cfi.pflash01,property=secure,value=on \
&gt;      -drive if=pflash,readonly=on,file=/usr/share/OVMF/OVMF_CODE.secboot.fd \
&gt;      -pflash win10-vars.img -hda win10-uefi.img -cdrom /vm/isos/en_windows_10_enterprise_n_version_1511_updated_apr_2016_x64_dvd_8709271.iso \
&gt;      -vga std -boot d -global isa-debugcon.iobase=0x402

Can you try again with Linux 4.6 and a command line as similar to the above
as possible?

**Comment by tianocore-issues** [comment]
## Comment 344
Date: 2016-07-21 16:24:47 +0000
From: @lersek

webczat commented on May 24

Well, I believe archlinux will have 4.6 in a short while and I will then
be able to try.

**Comment by tianocore-issues** [comment]
## Comment 345
Date: 2016-07-21 16:24:59 +0000
From: @lersek

webczat commented on May 24

ahh one thing to note. what does -smp 2 do? windows does not support
multiple physical processors. try -smp cores=2 unless you are sure that
it has the same effect. I mean the windows system will probably detect,
but not enable, the second cpu.

**Comment by tianocore-issues** [comment]
## Comment 346
Date: 2016-07-21 16:25:14 +0000
From: @lersek

bonzini commented on May 24

Also, please capture from dmesg anything that starts with *** Guest State ***. I am particularly interested in the line starting with VMEntry: and the three after that. You can get that with 4.5 too, and please capture it two-three times to see whether it&apos;s always the same or it changes. Finally, please include the output of grep -r . /sys/module/kvm_intel/parameters/ here in the bug.
